google/cloud/storage/client.h

```diff
@@ -533,6 +533,40 @@ class Client {
     return raw_client_->PatchObject(request).second;
   }
 
+  /**
+   * Compose existing objects into a new object in the same bucket.
+   *
+   * @param bucket_name the name of the bucket that contains the object.
+   * @param destination_object_name the composed object name.
+   * @param source_objects objects used to compose `destination_object_name`.
+   * @param destination_object_metadata the new metadata for the Object. Only
+   *     the writeable fields accepted by the `Objects: compose` API are used,
+   *     all other fields are ignored.
+   * @param options a list of optional query parameters and/or request headers.
+   *     Valid types for this operation include `ComposeObjectRequest`,
+   *      `EncryptionKey`, `Generation`, `IfGenerationMatch`,
+   *      `IfMetagenerationMatch`, `PredefinedAcl`,
+   *      `SourceEncryptionKey`, `UserProject`.
+   *
+   * @throw std::runtime_error if the operation fails.
+   *
+   * @par Example
+   * @snippet storage_object_samples.cc update object metadata
+   */
+  template <typename... Options>
+  ObjectMetadata ComposeObject(std::string bucket_name,
+                        std::string destination_object_name,
+                        std::vector<ComposeSourceObject> const& source_objects,
```

Comment:
You should pass this parameter by value and `std::move` it, you are making one extra copy otherwise.

Reply:
(Trying to understand your comment from reading a bit)
If I use `const&`, a copy of `source_objects` for the method `ComposeObject` I will have an additional copy.
Instead if I pass-by-value `std::vector<ComposeSourceObject> source_objects`, and use `std::move(calling_code_source_objects)` to move pointer addresses to the method parameter `source_objects` for `ComposeObject`.

This would save a copy, I think?

Reply:
I think your assessment is correct.  Looking at this closer: do you need to make a copy of this vector in `ComposeObjectRequest`?  If not, then we can just pass by const reference through both calls.

Reply:
Did a little more research.

Original purpose was to create a copy of the `source_objects` in the request to print out to `ostream`. 

At the moment I'm passing it by reference and creating a copy when constructing ComposeObjectRequest. I thought it might be useful to keep `source_objects` for `ostream`, but I'm using more memory for little benefit as I would have the information in `json_payload()` already. 

I can change this to not have that extra copy by passing by reference to construct the `ComposeObjectRequest` instance only and not keep a copy for `ostream`

Reply:
Thanks for checking.

> I can change this to not have that extra copy by passing by reference to construct the `ComposeObjectRequest` instance only and not keep a copy for `ostream`

Let's do that.